pkg/ebpf/c/tracer-ebpf.c

```diff
@@ -845,6 +906,176 @@ int kprobe__tcp_v4_destroy_sock(struct pt_regs* ctx) {
     return 0;
 }
 
+SEC("kprobe/udp_destroy_sock")
+int kprobe__udp_destroy_sock(struct pt_regs* ctx) {
+    struct sock* sk = (struct sock*)PT_REGS_PARM1(ctx);
+
+    if (sk == NULL) {
+        log_debug("ERR(udp_destroy_sock): socket is null \n");
+        return 0;
+    }
+
+    // get tracer status and return early if we are still offset-guessing
+    u64 zero = 0;
+    tracer_status_t* status = bpf_map_lookup_elem(&tracer_status, &zero);
+    if (status == NULL) {
+        return 0;
+    }
+
+    // get the port for the current sock
+    __u16 lport = 0;
+    bpf_probe_read(&lport, sizeof(lport), ((char*)sk) + status->offset_sport);
+    lport = ntohs(lport);
+
+    if (lport == 0) {
+        log_debug("ERR(udp_destroy_sock): lport is 0 \n");
+        return 0;
+    }
+
+    // decide if the port is bound, if not, do nothing
+    __u8* state = bpf_map_lookup_elem(&udp_port_bindings, &lport);
+
+    if (state == NULL) {
+        log_debug("kprobe/udp_destroy_sock: sock was not listening, will drop event");
+        return 0;
+    }
+
+    // set the state to closed
+    __u8 new_state = PORT_CLOSED;
+    bpf_map_update_elem(&udp_port_bindings, &lport, &new_state, BPF_ANY);
+
+    log_debug("kprobe/udp_destroy_sock: port %d marked as closed", lport);
+
+    return 0;
+}
+
+SEC("kprobe/sys_bind")
+int kprobe__sys_bind(struct pt_regs* ctx) {
+
+    __s32 fd = PT_REGS_PARM1(ctx);
+    __u64 tid = bpf_get_current_pid_tgid();
+
+    // determine if the fd for this process is an unbound UDP socket
+    __u64 fd_and_tid = (tid << 32) | fd;
+    __u64* u = bpf_map_lookup_elem(&unbound_sockets, &fd_and_tid);
+
+    if (u == NULL) {
+        log_debug("kprobe/sys_bind: bind happened, but not on a UDP socket\n");
+        return 0;
+    }
+
+    struct sockaddr* addr = (struct sockaddr*)PT_REGS_PARM2(ctx);
+    if (addr == NULL) {
+        return 0;
+    }
+
+    // sockaddr is part of the syscall ABI, so we can hardcode the offset of 2 to find the port.
+    u16 sin_port = 0;
+    bpf_probe_read(&sin_port, sizeof(u16), (char*)addr + 2);
+    sin_port = ntohs(sin_port);
+
+    // write to pending_binds to the retprobe knows we can mark this as binding.
+    bind_syscall_args_t args = {};
+    args.fd = fd;
+    args.port = sin_port;
+
+    bpf_map_update_elem(&pending_bind, &tid, &args, BPF_ANY);
+    log_debug("kprobe/sys_bind: started a bind on UDP port=%d fd=%d tid=%d\n", sin_port, fd, tid);
+
+
+
+    return 0;
+}
+
+SEC("kretprobe/sys_bind")
+int kretprobe__sys_bind(struct pt_regs* ctx) {
+  __u64 tid = bpf_get_current_pid_tgid();
+  int ret = PT_REGS_RC(ctx);
+
+  // bail if this bind() is not one we're instrumenting
+  bind_syscall_args_t *args;
+  args = bpf_map_lookup_elem(&pending_bind, &tid);
```

Comment:
Do we need to remove from `pending_bind` if the value was present?

Reply:
No, it's allowed to bind to a socket multiple times